### ISSUE TITLE
Research how to implement the city layout from TestVille

### ISSUE THREAD

**Issue by Richargh** [MaibornWolff/codecharta#497]
[TestVille](https://github.com/MaibornWolff/TestVille) uses a different building layout than Codecharta. We want to implement this street layout as an alternative to our squarified layout in CC. It can be turned on at the users request.

We need to research (not build) how to do the street layout for a folder hierarchy that is deeply nested. Supposedly TestVille can only go three layers deep. 

In a following story we'll build it.

**Comment by alschmut** [comment]
If I get it right, this issue arose from the question how to implement a good visualization for edges. And probably everybody will agree with this: to visualise actual edges (lines between nodes), there needs to be some space between the nodes to recognise the edges.

So in my opinion the idea of separating the map into smaller pieces/cluster makes sense. These cluster need to be automatically generated. Lets say any map will always be separated into 5-15 cluster. A cluster contains nodes, that have a short path distance to each other (to kinda keep the hierarchical structure).

The cluster get visualised as 3d-codemaps and spread all around the canvas.
Now as there is lots of space between the cluster, we can show edges (as lines/arrows on the ground/floor). Some of those might be inside a cluster and less visible, but thats okey, as those nodes have a close "path relationship". The more important edges are those between clusters.

This visualization ...
- needs to recognise clusters with nodes that are closely structured inside the directory-structure
- needs to manage, where to draw the cluster to match the linked edges (e.x. the most linked cluster in the middle)
- provides an easy solution for deep hierarchies -> multiple 3d-codemaps (already existing code)
- needs to connect the different codeMaps to link the edges from one to the other node between clusters

I was thinking about something like this:
<img width="100%" src="https://user-images.githubusercontent.com/23529226/59809453-8d333980-92f8-11e9-9a0f-f23e9bd38fcb.jpg">
<img width="50%" src="https://user-images.githubusercontent.com/23529226/59809452-8d333980-92f8-11e9-9ba7-08aa1d3b0f98.jpg"><img width="50%" src="https://user-images.githubusercontent.com/23529226/59809456-8dcbd000-92f8-11e9-8eb8-aa83c2c6fe6c.jpg">

**Comment by Richargh** [comment]
Thanks for putting in the time and adding the cool screenshots. I also did some research in #578 concerning these edges.

I love the `pieces/cluster` idea but the visualization already shows the problems. Which of these edges are important? CodeCharta provides a great big picture view for metrics and intuitively shows that a user should look at the **big, tall, red building**. We need that same information for edges. We need to provide the user some slider to show only the important edges. And when he opens CC he immediately sees a big picture view of the important edges.

**Comment by alschmut** [comment]
1. I think we need to separate the ways how nodes and edges work and are recognised as being "bad".
2. We are always trying to change source code (which is represented as a node). Edges are not *physically* existing. So even when looking at edges, we are trying to find the node, that has the wrong amount of edges.
3. Edges can be recognised only when having an overview of all edges. That means it really depends. And as always we need to check the source code to actually know what happening in there.

- Regarding the 1st screenshot:
4. the cluster in the middle is connected to many other clusters, which is okay, as long as the cluster isn't linked too often. If there were "too many" edges it probably means, this cluster should be split.
5. the lower clusters are connected to each other with just 2 edges. This might indicate a circle with an unwanted dependency

> preserve the structure from the normal codeview, don't change it when edges significantly change.

The structure will be similar, as the clusters are only generated by the hierarchical structure. That means the same cluster will exist, but might be shown on another spot on the canvas, as the spot will depend on the edges to visualise centric clusters with many edges in the middle.

> needs to recognise clusters with nodes that are closely structured inside the directory-structure: agreed, but how?

Somehow walking through the codemap and adding directories to a cluster, until we have the amount of cluster...

> provides an easy solution for deep hierarchies -> multiple 3d-codemaps (already existing code): what does this mean

That means that we will visualise multiple code maps (currently CC visualises one codemap)

**Comment by Richargh** [comment]
> 2. We are always trying to change source code (which is represented as a node). Edges are not physically existing.

That depends on the edge. An import is also an edge. 

> 4. If there were "too many" edges it probably means, this cluster should be split.

too many is hard to define, which is why I want this power in users hand to finetune the shown edges. 

* If a cluster depends on the API from another cluster, that's usually ok. If every building in that cluster depends on the API from another cluster, it's usually bad.
* If a building depends on a lot of other stuff, it's also usually bad.

> 5. the lower clusters are connected to each other with just 2 edges. This might indicate a circle with an unwanted dependency

Indicate a circle. That is probably the best feature to start with. Show all circles. Can we draw dashed lines as well?

> might be shown on another spot on the canvas

That is a big complaint companies have towards CC. That the tree map is not ordered. We need to change that behavior.

> Somehow walking through the codemap and adding directories to a cluster, until we have the amount of cluster...

My computer theory is gone in that regard. Could be k-means but I have forgotton all of that.

> That means that we will visualise multiple code maps (currently CC visualises one codemap)

basically we need to change our treemap algorithm. That is something we will have to do at one point to get an ordered treemap. Especially if we want to visualize multiple codemaps.

**Comment by Richargh** [comment]
@alschmut I just added #529 which is the most minimal edge improvement I could think of. It makes the edges visible and hopefully accessible. Do you want to leave comments in that issue?

After getting some user feedback form #529 we'll see which ideas from this issue might be helpful to our users.

**Comment by Richargh** [comment]
We'll take this issue as an inspiration for #788
